Failed to infer actual type parameters

When removing fixed from Sequence.type.equality
```
/home/not_synced/fuzion (2024-10-25T10-20-02+02-00)2$ git diff  | cat
diff --git a/lib/Sequence.fz b/lib/Sequence.fz
index eea70968b..1b027cc07 100644
--- a/lib/Sequence.fz
+++ b/lib/Sequence.fz
@@ -902,7 +902,7 @@ public Sequence(public T type) ref : property.equatable is
 
   # equality
   #
-  public fixed redef type.equality(a, b Sequence T) =>
+  public redef type.equality(a, b Sequence.this) =>
     if T : property.equatable
       a.count = b.count
         && ((a.zip b x,y->x=y) ∀ x->x)
/home/not_synced/fuzion (2024-10-25T10-20-02+02-00)$ make no-java 

./build/lib/Sequence.fz:908:16: error 1: Failed to infer actual type parameters
        && ((a.zip b x,y->x=y) ∀ x->x)

In call to 'Sequence.zip', no actual type parameters are given and inference of the type parameters failed.
Expected type parameters: 'U, V'
Type inference failed for 2 type parameters 'U', 'V'

one error.
make: *** [Makefile:605: build/modules/base.fum] Error 1
```
When trying to specify the types explicitly:
```
/home/not_synced/fuzion (2024-10-25T10-20-02+02-00)2$ git diff  | cat
diff --git a/lib/Sequence.fz b/lib/Sequence.fz
index eea70968b..e01dc3feb 100644
--- a/lib/Sequence.fz
+++ b/lib/Sequence.fz
@@ -902,10 +902,10 @@ public Sequence(public T type) ref : property.equatable is
 
   # equality
   #
-  public fixed redef type.equality(a, b Sequence T) =>
+  public redef type.equality(a, b Sequence.this) =>
     if T : property.equatable
       a.count = b.count
-        && ((a.zip b x,y->x=y) ∀ x->x)
+        && ((a.zip T bool b x,y->x=y) ∀ x->x)
     else
       panic "***"
 
/home/not_synced/fuzion (2024-10-25T10-20-02+02-00)$ make no-java 

./build/lib/Sequence.fz:908:27: error 1: Incompatible types when passing argument in a call
        && ((a.zip T bool b x,y->x=y) ∀ x->x)

Actual type for argument #1 'b' does not match expected type.
In call to          : 'Sequence.zip'
expected formal type: 'Sequence Sequence.type.T'
actual type found   : 'Sequence.this.type (in type feature)'
assignable to       : 'Sequence.this.type (in type feature)'
for value assigned  : 'b'
To solve this, you could change the type of the target 'b' to 'Sequence.this.type (in type feature)' or convert the type of the assigned value to 'Sequence Sequence.type.T'.

one error.
make: *** [Makefile:605: build/modules/base.fum] Error 1
```

--- a/modules/base/src/Sequence.fz
+++ b/modules/base/src/Sequence.fz
@@ -1521,8 +1521,8 @@ is
 
   # equality
   #
-  public fixed redef type.equality(a, b equatable_sequence T) =>
-    equality0(a1, b1 equatable_sequence (T2 : property.equatable)) =>
+  public redef type.equality(a, b equatable_sequence.this) =>
+    equality0(a1, b1 equatable_sequence.this) =>
       (a1.unwrap.zip b1.unwrap x,y->x=y) ∀ id
 
     au := a.unwrap

--- a/modules/base/src/array2.fz
+++ b/modules/base/src/array2.fz
@@ -36,7 +36,7 @@ private:public array2(
        redef internal_array fuzion.sys.internal_array T,
        _ unit,
        _ unit)
- : array T /* NYI: BUG: when removing T Failed to infer actual type parameters */ internal_array unit unit unit
+ : array internal_array unit unit unit
   pre
     safety: length0 ≥ 0
     safety: length1 ≥ 0

--- a/modules/base/src/array3.fz
+++ b/modules/base/src/array3.fz
@@ -36,7 +36,7 @@ private:public array3(
        redef internal_array fuzion.sys.internal_array T,
        _ unit,
        _ unit)
- : array T /* NYI: BUG: when removing T Failed to infer actual type parameters */ internal_array unit unit unit
+ : array internal_array unit unit unit
   pre
     safety: length0 ≥ 0
     safety: length1 ≥ 0

--- a/modules/base/src/fuzion/runtime/contract_fault.fz
+++ b/modules/base/src/fuzion/runtime/contract_fault.fz
@@ -29,7 +29,7 @@
 public contract_fault (
   # the handler this effect uses to fail
   p ((String,String)) -> void
-  ) : eff.fallible (String,String) /* NYI: BUG: this type parameter is required, type inference results in NullPointException */ p
+  ) : eff.fallible p
 is
 
 

--- a/modules/base/src/fuzion/runtime/fault.fz
+++ b/modules/base/src/fuzion/runtime/fault.fz
@@ -28,7 +28,7 @@
 public fault (
   # the handler this effect uses to fail
   p ((String, String))->void
-  ) : eff.fallible (String, String) /* NYI: BUG: this type parameter is required, type inference results in NullPointerException */ p
+  ) : eff.fallible p
 is
 
   # install default fault handler

--- a/modules/base/src/fuzion/runtime/post_fault.fz
+++ b/modules/base/src/fuzion/runtime/post_fault.fz
@@ -29,7 +29,7 @@
 public post_fault (
   # the handler this effect uses to fail
   p String -> void
-  ) : eff.fallible String /* NYI: BUG: this type parameter is required, type inference results in NullPointerException */ p
+  ) : eff.fallible p
 is
 
 

--- a/src/dev/flang/ast/Resolution.java
+++ b/src/dev/flang/ast/Resolution.java
@@ -425,6 +425,10 @@ public class Resolution extends ANY
 
         Feature f = forType.removeFirst();
         if (DEBUG) sayDebug("resolve types: " + f);
+        if (f.isCotype())
+          {
+            resolveTypes(f.cotypeOrigin());
+          }
         f.internalResolveTypes(this);
       }
     else if (!moreThanTypes)

--- a/src/dev/flang/ast/SrcModule.java
+++ b/src/dev/flang/ast/SrcModule.java
@@ -72,9 +72,9 @@ public interface SrcModule extends AbstractModule
                              boolean ignoreNotFound);
 
   void addCotype(AbstractFeature outerType,
-                      Feature         innerType);
+                 Feature         innerType);
   void addTypeParameter(AbstractFeature outerType,
-                      Feature         innerType);
+                        Feature         innerType);
 
   /*----------------------  methods needed by AIR  ----------------------*/
 

--- a/src/dev/flang/fe/SourceModule.java
+++ b/src/dev/flang/fe/SourceModule.java
@@ -675,9 +675,14 @@ part of the (((inner features))) declarations of the corresponding
   {
     findDeclarations(cotype, outerType);
     addDeclared(outerType, cotype);
-    cotype.scheduleForResolution(_res);
     resolveDeclarations(cotype);
   }
+
+
+  /**
+   * During type resolution, add a type parameter created for a free type like
+   * {@code T} in {@code f(x T) is ...}.
+   */
   public void addTypeParameter(AbstractFeature outer,
                                Feature typeParameter)
   {

